### ISSUE TITLE
Render flicker on first load, on production

### ISSUE THREAD

**Issue by donedgardo** [Hashnode/mern-starter#276]
Live example: https://jay-mern-test.herokuapp.com/
I thought it was becouse of heroku, 
but running it locally and in production seem to have the same problem.

Looking at this closed issue #226 it seems to be the same issue.

**Comment by jgautheron** [comment]
@frankleng any pointer as how to do that with this boilerplate?

**Comment by jgautheron** [comment]
In `client/index.js`:

```javascript
match({ routes: getRoutes(store), location }, (error, redirectLocation, renderProps) => {
  render(
    <AppContainer>
      <Provider store={store}>
        <IntlWrapper>
          <Router history={browserHistory} {...renderProps}>
            {getRoutes(store)}
          </Router>
        </IntlWrapper>
      </Provider>
    </AppContainer>,
    mountApp
  );
});
```

**Comment by karolgorecki** [comment]
The minimal change that worked for me is to change:

In index.js
```
const r = (err, redirect, props) => render(
  <AppContainer>
    <App store={store} p={props}/>
  </AppContainer>,
  mountApp
);

match({ routes, history: browserHistory }, r);
```

In App.js
```
export default function App(props) {
  return (
    <Provider store={props.store}>
      <IntlWrapper>
        <Router {...props.p} />
      </IntlWrapper>
    </Provider>
  );
}
```
After this the FOUC disappeared, but I have still some issue with onEnter which is called twice: on server side + client side.

**Comment by quocble** [comment]
Thanks @karolgorecki for the fix.  Just fyi, it will break hot reloading because the reload of App.js will load routes multiple times. Here's the fix.

```
if (process.env.NODE_ENV === 'production') {
  const routes = require('./routes');

  const r = (err, redirect, props) => render(
    <AppContainer>
      <App store={store} p={props} />
    </AppContainer>,
    mountApp
  );
  match({ routes, history: browserHistory }, r);
} else {
  render(
    <AppContainer>
      <App store={store}/>
    </AppContainer>,
    mountApp
  );
}
```

```
export default function App(props) {
  const router = (process.env.NODE_ENV === 'production') ?  (<Router {...props.p} />)
          :(<Router history={browserHistory}>
              {routes}
            </Router>);  
  return (
    <Provider store={props.store}>
      <IntlWrapper>
        {router}
      </IntlWrapper>
    </Provider>
  );
}
```

**Comment by charlessmith259** [comment]
How much did you pay to deploy it on heroku, specially for the database?

What were the steps in a nutshell?

Thanks,

Charles S.

**Comment by InamTaj** [comment]
@quocble I tried your solution but HMR is still broken. Were you able to fix that?

**Comment by jtdaugh** [comment]
Thanks @quocble - this fixed the issue for me. Wish it were built into mern-starter.

One change for me - I had to do `import routes from './routes` vs require

Anothing I noticed (in case it helps someone here in the future) was once the double rendering/flicker went away, some elements that were previously rendered how I intended were now rendering weird in production. This was caused by validateDomNesting errors - fixing those made the page render as it did before.
